File: qbopt/optimize/canonical.py
"""Canonical forms, as LLVM's InstCombine puts them: compares with the constant on the right, no neutral terms.

Folding makes `1 sub v` whenever a compare's left operand becomes known, and
no machine compares an immediate against a register in that order. The
compare is swapped and every test reading its flags mirrored; a reader that
is not a test gets the constant as a copied value instead.
"""

from dataclasses import replace

from qbopt.model import ir
from qbopt.model import mir
from qbopt.analysis import ssa
from qbopt.analysis import consts


def compares(body: mir.MirBody) -> mir.MirBody:
    readers: dict[mir.Value, list[mir.Op]] = {}
    for block in body.blocks:
        for op in block.ops:
            for value in op.uses:
                readers.setdefault(value, []).append(op)
    swapped: set[mir.Value] = set()
    copied: set[int] = set()
    for block in body.blocks:
        for op in block.ops:
            if not _constant_left(op):
                continue
            tests = all(
                one.kind is mir.Kind.BRANCH and one.test in mir.MIRRORED for one in readers.get(op.defines[0], ())
            )
            if tests and not isinstance(op.args[1], mir.Const):
                swapped.add(op.defines[0])
            else:
                copied.add(id(op))
    if not swapped and not copied:
        return body

    values = tuple(ssa.values(body))
    serial = max((value.id for value in values), default=0)
    variable = max((value.variable for value in values), default=0)
    blocks = []
    for block in body.blocks:
        ops = []
        for op in block.ops:
            if _constant_left(op) and op.defines[0] in swapped:
                ops.append(replace(op, args=(op.args[1], op.args[0])))
            elif id(op) in copied:
                serial += 1
                variable += 1
                held = mir.Value(serial, op.at, variable=variable, version=1)
                constant = op.args[0]
                ops.append(
                    mir.Op(
                        op.at,
                        ir.Operation.MOVE,
                        "",
                        (held,),
                        (),
                        kind=mir.Kind.COPY,
                        args=(constant,),
                        results=(mir.Held(held, constant.width),),
                    )
                )
                ops.append(replace(op, uses=(held, *op.uses), args=(mir.Held(held, constant.width), *op.args[1:])))
            elif op.kind is mir.Kind.BRANCH and swapped.intersection(op.uses):
                ops.append(replace(op, test=mir.MIRRORED[op.test]))
            else:
                ops.append(op)
        blocks.append(replace(block, ops=tuple(ops)))
    return replace(body, blocks=tuple(blocks))


def _constant_left(op: mir.Op) -> bool:
    return (
        op.kind is mir.Kind.SUB
        and not op.results
        and len(op.defines) == 1
        and len(op.args) == 2
        and isinstance(op.args[0], mir.Const)
    )


def identities(body: mir.MirBody) -> mir.MirBody:
    """`x + 0`, `x - 0` and `x * 1` are `x`, and a test `x <=u 0` is `x == 0`.

    A rewrite states what it computes from a proof in full -- rotation's
    trip count is `bound - start + inclusive` for any start -- and the
    neutral terms go here, so no rewrite folds its own.
    """
    read = {value for block in body.blocks for op in block.ops for value in op.uses}
    read |= {value for block in body.blocks for phi in block.phis for value in phi.incoming.values()}
    swap: dict[int, mir.Value] = {}
    copies: dict[int, mir.Op] = {}
    zero_tests: set[mir.Value] = set()
    for block in body.blocks:
        for op in block.ops:
            if _zero_test(op):
                zero_tests.add(op.defines[0])
            kept = _neutral(op)
            if kept is None or any(value.flags and value in read for value in op.defines):
                continue
            (result,) = op.results
            if isinstance(kept, mir.Held):
                swap[result.value.id] = kept.value
            else:
                copies[id(op)] = replace(
                    op,
                    kind=mir.Kind.COPY,
                    defines=(result.value,),
                    uses=(),
                    args=(kept,),
                    source_backed=False,
                    raised=None,
                    symbol=False,
                )
    tests = {mir.Kind.BELOW_EQ: mir.Kind.EQ, mir.Kind.ABOVE: mir.Kind.NE}
    renamed = any(
        op.kind is mir.Kind.BRANCH and op.test in tests and zero_tests.intersection(op.uses)
        for block in body.blocks
        for op in block.ops
    )
    if not swap and not copies and not renamed:
        return body
    blocks = []
    for block in body.blocks:
        ops = []
        for op in block.ops:
            if op.results and isinstance(op.results[0], mir.Held) and op.results[0].value.id in swap:
                if op.id is not None or op.source_backed or op.absorbed:
                    ops.append(mir.cleared(op))  # its bytes stay owned
                continue
            op = ssa.substituted(copies.get(id(op), op), swap)
            if op.kind is mir.Kind.BRANCH and op.test in tests and zero_tests.intersection(op.uses):
                op = replace(op, test=tests[op.test])  # nothing is below zero
            ops.append(op)
        phis = tuple(
            replace(phi, incoming={at: ssa.provider(value, swap) for at, value in phi.incoming.items()})
            for phi in block.phis
        )
        blocks.append(replace(block, phis=phis, ops=tuple(ops)))
    return replace(body, blocks=tuple(blocks))


def _neutral(op: mir.Op) -> mir.Held | mir.Const | None:
    """The operand a pure `x + 0`, `x - 0` or `x * 1` passes through unchanged."""
    if (
        len(op.args) != 2
        or len(op.results) != 1
        or not isinstance(op.results[0], mir.Held)
        or op.loads
        or op.stores
        or op.merges
        or op.barrier
        or op.floating is not None
    ):
        return None
    width = op.results[0].width
    identity = {mir.Kind.ADD: 0, mir.Kind.SUB: 0, mir.Kind.MUL: 1}.get(op.kind)
    if identity is None:
        return None
    left, right = op.args
    commutes = op.kind is not mir.Kind.SUB
    for kept, other in ((left, right), (right, left))[: 1 + commutes]:
        if (
            isinstance(other, mir.Const)
            and consts.masked(other.n, width) == identity
            and isinstance(kept, (mir.Held, mir.Const))
            and kept.width == width
        ):
            return kept
    return None


def _zero_test(op: mir.Op) -> bool:
    return (
        op.kind is mir.Kind.SUB
        and not op.results
        and len(op.defines) == 1
        and len(op.args) == 2
        and isinstance(op.args[1], mir.Const)
        and consts.masked(op.args[1].n, op.args[1].width) == 0
    )

File: crates/llrm-core/src/optimize/canonical.rs
//! Canonical forms, as LLVM's InstCombine puts them: compares with the
//! constant on the right, no neutral terms.
//!
//! Port of `qbopt/optimize/canonical.py`. Folding makes `1 sub v` whenever a
//! compare's left operand becomes known, and no machine compares an
//! immediate against a register in that order. The compare is swapped and
//! every test reading its flags mirrored; a reader that is not a test gets
//! the constant as a copied value instead.

use std::rc::Rc;
use std::collections::{BTreeMap, BTreeSet};

use crate::support::hash::IndexMap;

use crate::analysis::{consts, ssa};
use crate::model::ir::Operation;
use crate::model::mir::{self, Arg, Held, Kind, MirBody, Op, OpCode, Value};

pub fn compares(body: Rc<MirBody>) -> Rc<MirBody> {
    let mut readers = IndexMap::<Value, Vec<&Op>>::default();
    for block in &body.blocks {
        for op in &block.ops {
            for value in &op.uses {
                readers.entry(*value).or_default().push(op);
            }
        }
    }
    let mut swapped = BTreeSet::<Value>::new();
    // `id(op)`: the block and operation index.
    let mut copied = BTreeSet::<(usize, usize)>::new();
    for (b, block) in body.blocks.iter().enumerate() {
        for (i, op) in block.ops.iter().enumerate() {
            if !_constant_left(op) {
                continue;
            }
            let tests = readers.get(&op.defines[0]).map_or(&[][..], Vec::as_slice).iter().all(|one| {
                one.kind == Kind::Branch && one.test.is_some_and(|test| mir::MIRRORED(test).is_some())
            });
            if tests && !matches!(op.args[1], Arg::Const(_)) {
                swapped.insert(op.defines[0]);
            } else {
                copied.insert((b, i));
            }
        }
    }
    if swapped.is_empty() && copied.is_empty() {
        return body;
    }

    let values = ssa::values(&body).collect::<Vec<_>>();
    let mut serial = values.iter().map(|value| value.id).max().unwrap_or(0);
    let mut variable = values.iter().map(|value| value.variable).max().unwrap_or(0);
    let mut blocks = Vec::new();
    for (b, block) in body.blocks.iter().enumerate() {
        let mut ops = Vec::new();
        for (i, op) in block.ops.iter().enumerate() {
            if _constant_left(op) && swapped.contains(&op.defines[0]) {
                let mut changed = op.clone();
                changed.args = vec![op.args[1].clone(), op.args[0].clone()];
                ops.push(changed);
            } else if copied.contains(&(b, i)) {
                serial += 1;
                variable += 1;
                let held = Value { variable, version: 1, ..Value::new(serial, op.at) };
                let Arg::Const(constant) = &op.args[0] else {
                    unreachable!("a constant left operand");
                };
                let mut copy = Op::new(op.at, OpCode::Operation(Operation::Move), "", vec![held], Vec::new());
                copy.kind = Kind::Copy;
                copy.args = vec![Arg::Const(constant.clone())];
                copy.results = vec![Arg::Held(Held { value: held, width: constant.width })];
                ops.push(copy);
                let mut changed = op.clone();
                changed.uses = std::iter::once(held).chain(op.uses.iter().copied()).collect();
                changed.args = std::iter::once(Arg::Held(Held { value: held, width: constant.width }))
                    .chain(op.args[1..].iter().cloned())
                    .collect();
                ops.push(changed);
            } else if op.kind == Kind::Branch && op.uses.iter().any(|value| swapped.contains(value)) {
                let mut changed = op.clone();
                changed.test = op.test.map(|test| mir::MIRRORED(test).expect("KeyError: a mirrored test"));
                ops.push(changed);
            } else {
                ops.push(op.clone());
            }
        }
        blocks.push(block.with_ops(ops));
    }
    Rc::new(MirBody { blocks, ..MirBody::clone(&body) })
}

fn _constant_left(op: &Op) -> bool {
    op.kind == Kind::Sub
        && op.results.is_empty()
        && op.defines.len() == 1
        && op.args.len() == 2
        && matches!(op.args[0], Arg::Const(_))
}

/// `x + 0`, `x - 0` and `x * 1` are `x`, and a test `x <=u 0` is `x == 0`.
///
/// A rewrite states what it computes from a proof in full -- rotation's
/// trip count is `bound - start + inclusive` for any start -- and the
/// neutral terms go here, so no rewrite folds its own.
pub fn identities(body: Rc<MirBody>) -> Rc<MirBody> {
    let mut read = BTreeSet::<Value>::new();
    for block in &body.blocks {
        read.extend(block.ops.iter().flat_map(|op| op.uses.iter().copied()));
        read.extend(block.phis.iter().flat_map(|phi| phi.incoming.values().copied()));
    }
    let mut swap = BTreeMap::<u32, Value>::new();
    // `id(op)`: the block and operation index.
    let mut copies = BTreeMap::<(usize, usize), Op>::new();
    let mut zero_tests = BTreeSet::<Value>::new();
    for (b, block) in body.blocks.iter().enumerate() {
        for (i, op) in block.ops.iter().enumerate() {
            if _zero_test(op) {
                zero_tests.insert(op.defines[0]);
            }
            let Some(kept) = _neutral(op) else { continue };
            if op.defines.iter().any(|value| value.flags && read.contains(value)) {
                continue;
            }
            let Arg::Held(result) = &op.results[0] else { unreachable!("_neutral checked a held result") };
            if let Arg::Held(held) = &kept {
                swap.insert(result.value.id, held.value);
            } else {
                let mut copy = op.clone();
                copy.kind = Kind::Copy;
                copy.defines = vec![result.value];
                copy.uses = Vec::new();
                copy.args = vec![kept];
                copy.source_backed = false;
                copy.raised = None;
                copy.symbol = Some(false);
                copies.insert((b, i), copy);
            }
        }
    }
    let renamed_test = |op: &Op| {
        op.kind == Kind::Branch
            && matches!(op.test, Some(Kind::BelowEq | Kind::Above))
            && op.uses.iter().any(|value| zero_tests.contains(value))
    };
    let renamed = body.blocks.iter().flat_map(|block| &block.ops).any(|op| renamed_test(op));
    if swap.is_empty() && copies.is_empty() && !renamed {
        return body;
    }
    let mut blocks = Vec::new();
    for (b, block) in body.blocks.iter().enumerate() {
        let mut ops = Vec::new();
        for (i, op) in block.ops.iter().enumerate() {
            if let Some(Arg::Held(result)) = op.results.first() {
                if swap.contains_key(&result.value.id) {
                    if op.id.is_some() || op.source_backed || !op.absorbed.is_empty() {
                        ops.push(mir::cleared(op)); // its bytes stay owned
                    }
                    continue;
                }
            }
            let mut op = ssa::substituted(copies.get(&(b, i)).unwrap_or(op), &swap).expect("an acyclic substitution");
            if renamed_test(&op) {
                // Nothing is below zero.
                op.test = Some(if op.test == Some(Kind::BelowEq) { Kind::Eq } else { Kind::Ne });
            }
            ops.push(op);
        }
        let mut block = block.clone();
        for phi in &mut block.phis {
            phi.incoming = phi
                .incoming
                .iter()
                .map(|(at, value)| (*at, ssa::provider(*value, &swap).expect("an acyclic substitution")))
                .collect();
        }
        block.ops = ops;
        blocks.push(block);
    }
    Rc::new(MirBody { blocks, ..MirBody::clone(&body) })
}

/// The operand a pure `x + 0`, `x - 0` or `x * 1` passes through unchanged.
fn _neutral(op: &Op) -> Option<Arg> {
    if op.args.len() != 2
        || op.results.len() != 1
        || !matches!(op.results[0], Arg::Held(_))
        || !op.loads.is_empty()
        || !op.stores.is_empty()
        || !op.merges.is_empty()
        || op.barrier()
        || op.floating.is_some()
    {
        return None;
    }
    let Arg::Held(result) = &op.results[0] else { return None };
    let width = result.width;
    let identity = match op.kind {
        Kind::Add | Kind::Sub => 0,
        Kind::Mul => 1,
        _ => return None,
    };
    let (left, right) = (&op.args[0], &op.args[1]);
    let pairs = [(left, right), (right, left)];
    let commutes = op.kind != Kind::Sub;
    for (kept, other) in &pairs[..1 + usize::from(commutes)] {
        let Arg::Const(other) = other else { continue };
        if consts::masked(&other.n, width) != identity.into() {
            continue;
        }
        match kept {
            Arg::Held(held) if held.width == width => return Some((*kept).clone()),
            Arg::Const(constant) if constant.width == width => return Some((*kept).clone()),
            _ => {}
        }
    }
    None
}

fn _zero_test(op: &Op) -> bool {
    op.kind == Kind::Sub
        && op.results.is_empty()
        && op.defines.len() == 1
        && op.args.len() == 2
        && matches!(&op.args[1], Arg::Const(constant) if consts::masked(&constant.n, constant.width) == 0.into())
}

#[cfg(test)]
#[path = "canonical_tests.rs"]
mod tests;
